### DISCUSSION TITLE
Collaboration mode - Self-hosting vs Collab on top of <Excalidraw/>

### DISCUSSION THREAD

**Discussion by pw-web** [excalidraw/excalidraw#3879] [Q&A] [answered]
Hi team
Thanks for creating Excalidraw!

I am looking to integrate it in my React webapp. I used [Excalidraw](https://www.npmjs.com/package/@excalidraw/excalidraw) and it is working great, but I also need to implement collaboration mode.

**I looked at these options:**
1. We can implement own collab mode on top of the Excalidraw component - as suggested by Aakansha [here](https://discord.com/channels/723672430744174682/818853761886715964/824585589839757313)
2. We can self-host Excalidraw - as suggested [here](https://github.com/excalidraw/excalidraw/issues/1772)

**Issues:**
1. I will already be using WebSockets in my application for other functionalities. But developing own collab mode seems difficult, time-consuming, and prone to bugs, looking at comments of people who have tried it before.
2. Self-hosting causes a different problem for me. I don't need only Excalidraw on the page, I want to render it inside a <div> along with other components of my application.

How should I proceed? What would the team suggest me to do?
Thanks

**Comment by pw-web** [comment]
Follow-up question regarding developing collab on top of <Excalidraw/> component:

I was also reading Excalidraw's P2P collab [blog](https://blog.excalidraw.com/building-excalidraw-p2p-collaboration-feature/). I saw several operations being performed on collab data like adding, deleting, dealing with conflicts. And it was mentioned there is more to do.

1. Do we also have to take care of these if we are implementing collab on top of the <Excalidraw/> component?
2. Can we rely on CRDT libraries like [Automerge](https://github.com/automerge/automerge), [ShareDB](https://github.com/share/sharedb), [Yjs](https://github.com/yjs/yjs/), [Convergence](https://convergence.io/), etc, for collaborative data merging?

**Reply by ad1992** [reply]
We have our own [reconcilation algo](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/CollabWrapper.tsx#L387) which takes care of resolving merge conflicts.
> Do we also have to take care of these if we are implementing collab on top of the component?

Yes if you are implementing collaboration you will have to take care of it. 

> Can we rely on CRDT libraries like Automerge, ShareDB, Yjs, Convergence, etc, for collaborative data merging?

Since collaboration is something consumer is taking care of so its completely up to you how you want to implement it.  We trigger [`onChange`](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/README.md#onChange) callback whenever there is some change in excalidraw so there might be some additional data you might need along with `elements` and `appState` so that we can discuss based on the implementation.

**Comment by pw-web** [comment]
My specific use-case is to let one person draw and multiple other people can see (in view mode). Is there any easier way to achieve this?

Can we export the data from the writer, send it over to others via sockets, import in on the readers' side? Will that be fast enough for the users?

**Reply by Yoname** [reply]
Hi, @pw-web I release a WebSocket project that helps build multiplayer apps, we just release an integration of Excalidraw and multiplayer cursor chat. https://github.com/yomorun/presencejs  Hope this will help, and feel free to reach out  yona@yomo.run

**Comment by ad1992** [comment, marked as answer]
Hi @pw-web 👋🏻 thanks for the question!
As the npm package or self-host Excalidraw doesn't come with collaboration so the consumer needs to implement it themselves. 

As per your requirement you want to implement collaboration in excalidraw without writing from scratch. If you don't want to implement your own collaboration right now and reuse excalidraw's collaboration algorithm. 
You can try out the below steps 👇🏻 
1. Clone [excalidraw-room](https://github.com/excalidraw/excalidraw-room) and host the server yourself
2. Clone [excalidraw](https://github.com/excalidraw/excalidraw) and update the [server end point](https://github.com/excalidraw/excalidraw/blob/ac66665b6472c46d0674065872a7b30a705ac336/.env.production#L10) as well.
3. Update the [entry](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/webpack.prod.config.js#L12) to [excalidraw-app/index.tsx](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/index.tsx) so that the bundle contains the package with collaboration.
4. Release the npm package and start using it.

If you want to avoid steps 2-4 and have your own implementation, you can take a look at [collaboration](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/CollabWrapper.tsx) of how we do at excalidraw and reuse the same at your end so you don't have to clone the excalidraw repo and change end points etc. 

Though steps 2-4 will help you implement collab faster but I will suggest you avoid cloning excalidraw repo so you need not maintain your own npm package and rather copy the code which you want to use so you can modify it as well if needed.

Let me know if this helps!

We will soon try to make this process easier for consumers so its easy to use collaboration with excalidraw with your own servers.

## Update post ESM
The above steps would work if you are using the `umd` build but since we have moved to `esm` it will not work so  please follow the below steps for `ESM` build

Steps 1 and 2 remain the same as above
3. Update the entry points [here](https://github.com/excalidraw/excalidraw/blob/80ea7ca23f7cdeb12ecc73271470a501b2c4f7f8/scripts/buildPackage.js#L43) and [here](https://github.com/excalidraw/excalidraw/blob/80ea7ca23f7cdeb12ecc73271470a501b2c4f7f8/scripts/buildPackage.js#L100) in [buildPackage.js](https://github.com/excalidraw/excalidraw/blob/master/scripts/buildPackage.js#L100)  to [../excalidraw-app/App.tsx](https://github.com/excalidraw/excalidraw/blob/master/excalidraw-app/App.tsx) so that the bundle contains the package with collaboration.
4. Run the script -> `node scripts/release.js` and pass the release version, this will build and publish the package as well

**Reply by pw-web** [reply]
Thanks for your suggestions @ad1992 

As suggested by you, I will try and implement collaboration on my own on top of the [Excalidraw](https://www.npmjs.com/package/@excalidraw/excalidraw) component. I will update about any progress, or resources that might help others in the future.

**Reply by d0verenny** [reply]
This hyperlink is broken in 
"3. Update the [entry](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/webpack.prod.config.js#L10) to [excalidraw-app/index.tsx](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/index.tsx) so that the bundle contains the package with collaboration."

Could you please elaborate on the "entry" part and what and where do we have to change?

**Reply by d0verenny** [reply]
The hyperlink in this part is broken as well: 
If you want to avoid steps 2-4 and have your own implementation, you can take a look at [collaboration](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/CollabWrapper.tsx) of how we do at excalidraw and reuse the same at your end so you don't have to clone the excalidraw repo and change end points etc.

**Reply by ad1992** [reply]
I meant [this](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/webpack.prod.config.js#L12) to should be updated to [excalidraw-app/index.tsx](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/index.tsx), the line was updated hence its showing wrong, updated now.

**Reply by d0verenny** [reply]
Thank you for the reply! 
Could you please elaborate, in what way this should be updated? I'm so far from coding and scripting yet. 
Did you mean that I should replace `"excalidraw.production.min": "./entry.js", ` with `excalidraw-app/index.tsx` ? 
What is the correct way to do that without breaking the syntax (quotes, commas, and all that)?

And question No.2 regarding your point 2: "Clone [excalidraw](https://github.com/excalidraw/excalidraw) and update the [server end point](https://github.com/excalidraw/excalidraw/blob/master/.env.production#L4) as well."

The hyperlink on "server end point" — again — seem to be broken. It shows the string with the library part. 
Did you mean that user's own host address should be entered here: 
REACT_APP_WS_SERVER_URL=
Or did you mean here: 
REACT_APP_PORTAL_URL=https://portal.excalidraw.com
(replace https://portal.excalidraw.com with user's host address — e.g. http://localhost:5001)
?

**One more thing!** 
I'd like to also note that I'd like to later pack this as an image for docker. I'm saying this just to make sure if some of these steps are not needed for that.

Thank you very much for the help! There is no proper documentation and your little guide seem to be the only source :(

**Reply by ad1992** [reply]
> Thank you for the reply!
Could you please elaborate, in what way this should be updated? I'm so far from coding and scripting yet.
Did you mean that I should replace "excalidraw.production.min": "./entry.js", with excalidraw-app/index.tsx ?
What is the correct way to do that without breaking the syntax (quotes, commas, and all that)?

You need to replace "./entry.js" with  "../src/excalidraw-app/index.js" so it pickes the `index.js` file inside `excalidraw-app` as entry point.
For  updating hostm I meant `REACT_APP_WS_SERVER_URL`, updated above as well. Thanks for pointing out.

We are soon migrating to better docs so this things will be better maintained once migrated.

> I'd like to also note that I'd like to later pack this as an image for docker. I'm saying this just to make sure if some of these steps are not needed for that.

Currently collaboration is not supported in docker so we would need help in supporting for docker. In case someone wants to help they are welcome :)

**Reply by antwacky** [reply]
Hello, what needs to be done to support in docker?

**Reply by antwacky** [reply]
I got this working on docker, you just need to set REACT_APP_WS_SERVER_URL as an arg in the Dockerfile and then build the image (as well as doing step 3 mentioned in the original post).

I also have a working helm chart that will deploy excalidraw with collaboration. If anyone wants it, hook me up.

**Reply by d0verenny** [reply]
> I got this working on docker, you just need to set REACT_APP_WS_SERVER_URL as an arg in the Dockerfile and then build the image (as well as doing step 3 mentioned in the original post).
> 
> I also have a working helm chart that will deploy excalidraw with collaboration. If anyone wants it, hook me up.

Could you please elaborate:
1) how and where (in what line in the Docker file) exactly did you add REACT_APP_WS_SERVER_URL as an argument, and what does it look like?
2) Are you talking about Ecalidraw's Docker file or Ecalidraw Room's one? Thank you in advance.

**Reply by ryan-mcd** [reply]
> I got this working on docker, you just need to set REACT_APP_WS_SERVER_URL as an arg in the Dockerfile and then build the image (as well as doing step 3 mentioned in the original post).
> 
> I also have a working helm chart that will deploy excalidraw with collaboration. If anyone wants it, hook me up.

@UnionJoin  I would like to see your helm chart and container that enables collaboration. I couldn't find a related repo in your user account. If you just did it on  your local machine, I'll probably set up some CI to build this. However, the helm chart would be useful instead of building it myself.

**Reply by RudraSen2** [reply]
[This link](https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/CollabWrapper.tsx) is invalid. please update

**Reply by antwacky** [reply]
> > I got this working on docker, you just need to set REACT_APP_WS_SERVER_URL as an arg in the Dockerfile and then build the image (as well as doing step 3 mentioned in the original post).
> 
> > 
> 
> > I also have a working helm chart that will deploy excalidraw with collaboration. If anyone wants it, hook me up.
> 
> 
> 
> @UnionJoin  I would like to see your helm chart and container that enables collaboration. I couldn't find a related repo in your user account. If you just did it on  your local machine, I'll probably set up some CI to build this. However, the helm chart would be useful instead of building it myself.

I'll post this in the next week or so.

**Reply by JonasJR** [reply]
This would really be interesting!! Hope to see that helm chart soon @antwacky

**Reply by ranoble** [reply]
Double thumbs up on this @antwacky, also looking forward to taking a gander.

**Reply by ryan-mcd** [reply]
I got tired of waiting, and decided to produce this myself.

* You can obtain the image via ghcr at [https://github.com/ryan-mcd/containers/pkgs/container/excalidraw-collab](https://github.com/ryan-mcd/containers/pkgs/container/excalidraw-collab).
* I don't use kubernetes manifests directly; I use flux to deploy bjw-s's app-template helm chart: [https://github.com/bjw-s/helm-charts](https://github.com/bjw-s/helm-charts). You can find my example at [https://github.com/ryan-mcd/gitops-home-cluster/tree/main/cluster/apps/home/excalidraw/app](https://github.com/ryan-mcd/gitops-home-cluster/tree/main/cluster/apps/home/excalidraw/app).
* The environment variable that needs set is COLLAB_ADDR. Your excalidraw server instance needs set via COLLAB_ADDR=<Your Collab Server Address> e.g. COLLAB_ADDR=https://draw-collab.mydomain.tld
* I haven't confirmed, but I believe to deploy this, both excalidraw-room and excalidraw need to have separate URLs or subdomains and a valid certificate. I don't believe subpath will work. It may be possible to have both of these run unencrypted as well, but I have not tested this.

**Reply by PseudoResonance** [reply]
To put Excalidraw in a subdirectory, you can define a base path in `vite.config.ts`. Example:
```js
export default defineConfig({
  base: '/excalidraw/',
  server: {
```
As for the collab server, SocketIO strips the path for some reason, so you have to define a path where it's initialized in `src/excalidraw-app/collab/Collab.tsx` The path must end in `/socket.io/` as that is the endpoint used by the server. Example:
```js
        socketIOClient(socketServerData.url, {
          path: '/excalidrawcollab/socket.io/',
          transports: socketServerData.polling
```
I then used Nginx to strip the `/excalidrawcollab/` part of the location with the following, the `/` after proxy_pass and the Upgrade/Connection headers being the most important.
```nginx
        location /excalidrawcollab/ {
                proxy_pass http://host:port/;
                ...
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection $connection_upgrade;
        }
```

I think it would be good if both the server and client had proper environment variables to allow you to set the expected path as sometimes a subdomain isn't a viable option.

I've made the necessary changes to allow for these to be set in environment variables excalidraw/excalidraw#6990, excalidraw/excalidraw-room#363

I was also wondering about the Firebase server, as with this configuration, it is still dependent on Excalidraw's public Firebase to store any drawing data, so there's really little point in "self-hosting". Looking into it, if self hosting is important enough of a feature, there are services such as AppWrite or Supabase that claim to be mostly compatible with Firebase while offering native support for self hosting. Obviously that would be a huge ask though, as Excalidraw would either have to support both, or migrate their official hosting to an alternative.

Edit: I just realized there are some issues about this and it's been suggested to abstract away Firebase to allow for a variety of storage backends.

**Reply by itsnitinr** [reply]
Hi, is the original answer by @ad1992 still relevant? Most of the hyperlinks are broken. Seems like the codebase has changed by quite a bit. An update to this answer would be very helpful.

**Reply by ad1992** [reply]
@itsnitinr I just updated the post, can you try now and let me know if it works for you ?

**Reply by itsnitinr** [reply]
Thank you @ad1992. Could you please clarify what value I need to update with in the 3rd step of post ESM. Also, is it possible to enable collaboration by default without the user having to manually trigger it via `<Excalidraw />` component? I see that there's an `isCollaborating` prop but does it actaually start a collaboration session? If yes, how do I pass the room id? Thanks again.

**Reply by ad1992** [reply]
I just updated the post again, it should be [../excalidraw-app/App.tsx](https://github.com/excalidraw/excalidraw/blob/master/excalidraw-app/App.tsx)

> Also, is it possible to enable collaboration by default without the user having to manually trigger it via <Excalidraw /> component? 

No, we don't push the npm package with collaboration hence it isn't possible. 
The `isCollaborating` prop is to determine whether the the app is in collaboration mode so it can render the collaborator avatars

**Reply by itsnitinr** [reply]
Thank you. If I understood correctly, the only way to possibly achieve it would be to fork excalidraw and change a bunch of stuff in packages/excalidraw? Are there any plans to expose `collabApi` similar to `excalidrawApi` to call `startCollaboration()`?

**Reply by ad1992** [reply]
Yes, you will have to fork the repo and do the changes as mentioned above. Additionally I have updated the release script hence step **4** is now simplified (updated the answer as well).

> Are there any plans to expose collabApi similar to excalidrawApi to call startCollaboration()

Not yet, but may be in future we might expose a wrapper for collaboration but unsure as of now.

**Reply by mbj36** [reply]
@ad1992  `../excalidraw-app/App.tsx` won;t work due to `cwd` be `packages/excalidraw`

correct entryPoints on [buildPackage.js](https://github.com/excalidraw/excalidraw/blob/master/scripts/buildPackage.js#L43) and line 100 should be this

```const browserConfig = {
  entryPoints: ["../../excalidraw-app/App.tsx"],
  bundle: true,
  format: "esm",
  plugins: [
    sassPlugin(),
    woff2BrowserPlugin(),
    externalGlobalPlugin({
      react: "React",
      "react-dom": "ReactDOM",
    }),
  ],
  splitting: true,
}
```

**Reply by mbj36** [reply]
@ad1992 I tried Step 3

Some of the details which i think it's missing are - 

- First step to publish after doing step 3 would be update packages/excalidraw/package.json main and entry file, which is not documented anywhere

- When you change the entryPoints to ../../excalidraw-app/App.tsx, it expects some more npm package to be installed, which is not present when you build the package

- excalidraw-app/App.tsx has some environment variables, for example FIREBASE_CONFIG, how are we supposed to pass it when exported as package ?

**Comment by akashg3627** [comment]
hey @ad1992 
I wanted to use excalidraw collaboration mode with an option to have save button that saves the current canvas into my backend.
I am trying to implement the same using 4 steps you mentioned. I needed some help and guidance on how do we publish the package as the folder lies deep inside. So, can you please guide me in how do we publish the package as in which folder should be taken as root folder?

**Reply by ad1992** [reply]
Hey @akashg3627  👋🏻, do you mind asking about publish package as a separate question?

**Reply by akashg3627** [reply]
Yeah sure, would be better

**Comment by finallyivn** [comment]
Why does my hosted excalidraw server only work on my own computer (using two different browsers) after completing points 1 and 2 above, but not on other computers? Points 3 and 4, I don't understand

**Reply by Joshua-onwuzu** [reply]
@ad1992 is there a step by step process of releasing it as an npm package, i. dont have experiences with releasing npm packages so i have been stucked with this for quite sometime now. i tried the scripts on the package.json (release, autorelease, prerelease) but none of them work for me,  please can you help me out

**Reply by ad1992** [reply]
Hey @Joshua-onwuzu I will add this to the docs as well, here are the steps

1. update the package name in https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/package.json
2. run [prerelease](https://github.com/excalidraw/excalidraw/blob/master/scripts/prerelease.js) eg 👇🏻 
```
node scripts/prerelease 0.15.0
```
3. run the [release](https://github.com/excalidraw/excalidraw/blob/master/scripts/release.js)

```
node scripts/release.js
```

Its done ✨

**Reply by Joshua-onwuzu** [reply]
@ad1992  thanks alot for getting back to me, i tried what you just said but get these errors when i try to run the release script even after running this ```npx update-browserslist-db@latest```

**Reply by ad1992** [reply]
looks like `yarn build:umd` is failing for you, could you try running just this command and check the error ? Make sure your dependencies are upto date in the root folder.

**Reply by Joshua-onwuzu** [reply]
@ad1992  i ran ```yarn build:umd```  in the packages/excalidraw folder and it seems it runs successfully 
<img width="1440" alt="Screenshot 2023-04-18 at 08 19 13" src="https://user-images.githubusercontent.com/77381596/232701686-7f285e95-21c5-48b8-a4b0-dace1728c48a.png">

**Reply by Joshua-onwuzu** [reply]
@ad1992  now i get this error when i try running  ```yarn prelease [version number]```
<img width="1222" alt="Screenshot 2023-04-18 at 08 47 44" src="https://user-images.githubusercontent.com/77381596/232708724-3b7035fe-0963-4897-a41b-e47ad2482cf4.png">

**Reply by ad1992** [reply]
@Joshua-onwuzu lets keep it simple then

After updating package name and version
```
yarn build:umd
yarn publish
```

**Reply by Joshua-onwuzu** [reply]
@ad1992  i recloned the package and followed the steps from afresh and i was able to release the package but i get this error  on the browser when i try to  import ```Excalidraw``` from the released package
```Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.```  , you got any idea where is might be coming from ??
<img width="1440" alt="Screenshot 2023-04-19 at 06 30 33" src="https://user-images.githubusercontent.com/77381596/232975703-9e77e81b-0b97-43a3-8b8d-61476f483664.png">



here is how i am importing it in my code
<img width="1226" alt="Screenshot 2023-04-19 at 06 34 00" src="https://user-images.githubusercontent.com/77381596/232976254-4306f09b-0da2-4be9-8552-b701c4ad9967.png">

**Reply by ad1992** [reply]
@Joshua-onwuzu this happens when components are not exported correctly, probably you might have made some changes in the code due to which comp wasn't exported correctly. If you can share a code sandbox example / changes you made, I might be able to check

**Reply by Joshua-onwuzu** [reply]
@ad1992  sure, here in the modified package on codesandbox 
https://codesandbox.io/p/github/Joshua-onwuzu/excalidraw/main?file=%2FREADME.md&workspace=%257B%2522activeFilepath%2522%253A%2522%252FREADME.md%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clgncgeie000x3b6naismdokl%2522%253A%257B%2522key%2522%253A%2522clgncgeie000x3b6naismdokl%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522key%2522%253A%2522clgncgeie000y3b6ne5jlz3sx%2522%252C%2522type%2522%253A%2522PROJECT_SETUP%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clgnchkgc00843b6nj4952ora%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522clgnchi6t004t3b6nf2pg9sqj%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clgncgeie000x3b6naismdokl%2522%252C%2522spacesOrder%2522%253A%255B%2522clgncgeie000x3b6naismdokl%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D

**Reply by Joshua-onwuzu** [reply]
@ad1992  i didnt edit the code that much, just the changes that were suggested above, so i am not sure of what that is not exported correctly

**Comment by prephyte** [comment]
could you perhaps add a docker env variable like `COLLAB_SERVER_URL`, so that this would be easier to implement?

**Reply by okanji** [reply]
That would be helpful

**Comment by archef2000** [comment]
@ad1992 Is there now an easy way to get collab working on the selfhosted version?

**Comment by Mihai-CMM** [comment]
Is this a dead topic? Are there any helm charts with all the requirements out there? The last docker image cannot be built and there is PR opened. And also the documentation on how we should be able to integrate the 2 seems outdated
Thank you

**Reply by Anddd7** [reply]
I tried https://github.com/alswl/excalidraw-collaboration, it works for me.

Base on alswl's changes, excalidraw hardcoded backend api, so you cannot enable collaboration without changing the code. :see_no_evil:

---

I deployed it in kubernetes, if you wanna try: https://github.com/Anddd7/rubber-duck/tree/main/excalidraw

**Comment by isif00** [comment]
Hello @ad1992.
1-I cloned excalidraw and updated "index.tsx" to "../../excalidraw-app/App.tsx" [here](https://github.com/excalidraw/excalidraw/blob/d99e4a23ca0d9a6162e6c78ff52646189bc41d1a/scripts/buildPackage.js#L42) and [here](https://github.com/excalidraw/excalidraw/blob/d99e4a23ca0d9a6162e6c78ff52646189bc41d1a/scripts/buildPackage.js#L116).
2- I updated [VITE_APP_WS_SERVER_URL](https://github.com/excalidraw/excalidraw/blob/d99e4a23ca0d9a6162e6c78ff52646189bc41d1a/.env.production#L13) to the host running [excalidraw-room](https://github.com/excalidraw/excalidraw-room).
3- I updated the package name [here](https://github.com/excalidraw/excalidraw/blob/d99e4a23ca0d9a6162e6c78ff52646189bc41d1a/packages/excalidraw/package.json#L2) as well.
4- after that I need to run `node scripts/release.js`
but I'm getting this problem as you can see in the screenshot.
![image](https://github.com/user-attachments/assets/97ff64b1-195e-4e43-b086-94d4c8d6d958)


well, what I'm trying to do after this is to publish the package, and then use it in my react app <Excalidraw />, and pass the list of collaborators, and find a way to pass a room id. is that's a good approach ? is it possible to do it ? 
for more context about what i'm trying to do: 
```
currently I'm working on a collaborative platform, where I have workspaces, and each workspace has some features, one of them is excalidraw board, I want to allow the members of the workspaces to have a real time collaboration on that board.
```


thanks y'all 🙏

**Comment by rhagstrand** [comment]
Hi! People seem to be having alot of issues with this. Can anyone confirm that they have actually succeeded with using Excalidraw integrated in another website WITH the collaboration-features working using excalidraw-room as server as instructed above? I just need to know that it is indeed possible before trying the above mentioned steps.

**Reply by mbj36** [reply]
Nope, We tried hard, we couldn't succed and it seems issues are not being looked now on

**Reply by beltebelt** [reply]
It is possible. If you look at the code, you can see that they have a wrapper around the core library. That wrapper comes with collaboration features, but it isn't part of the npm package.

All you need to do is to create a fork to publish excalidraw with collab feature.

I was also able make some changes so that we could plug in any self hosted document store/object storage backend, instead of being restricted to firebase.

[Examples](https://github.com/beltebelt/excalidraw-with-storage-examples/tree/main/src/examples)

**Reply by l4rm4nd** [reply]
Got it working using the official images. 

No custom built images. Just the official excalidraw and excalidraw-rooms images from Dockerhub.

https://github.com/excalidraw/excalidraw/issues/5188#issuecomment-3848252703

https://blog.lrvt.de/selfhosting-excalidraw-with-collaboration-support/

**Comment by matbgn** [comment]
Here a working configuration thanks to amazing job of @alswl :

```
services:
  frontend:
    restart: always
    image: alswl/excalidraw:v0.17.3-fork-b1
    environment:
      - VITE_APP_BACKEND_V2_GET_URL=${EXACLIDRAW_CUSTOM_URL}/api/v2/
      - VITE_APP_BACKEND_V2_POST_URL=${EXACLIDRAW_CUSTOM_URL}/api/v2/post/
      - VITE_APP_WS_SERVER_URL=${EXACLIDRAW_CUSTOM_URL}
      - VITE_APP_FIREBASE_CONFIG={}
      - VITE_APP_HTTP_STORAGE_BACKEND_URL=${EXACLIDRAW_CUSTOM_URL}/api/v2
      - VITE_APP_STORAGE_BACKEND=http
    ports:
      - 8880:80
    depends_on:
      - storage-db
  storage:
    image: alswl/excalidraw-storage-backend:v2023.11.11
    restart: always
    environment:
      # docs https://github.com/alswl/excalidraw-storage-backend#environement-variables
      - STORAGE_URI=postgresql://excalidraw:${POSTGRES_PASSWORD}@storage-db/excalidraw
    ports:
      - 8881:8080
  storage-db:
    image: postgres:16.2
    volumes:
      - storage-db-data:/var/lib/postgresql/data
    restart: always
    environment:
      POSTGRES_USER: excalidraw
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_DB: excalidraw
  room:
    restart: always
    image: excalidraw/excalidraw-room:sha-03ff435
    ports:
      - 8882:80
volumes:
  storage-db-data: {}
networks: {}
```

**Comment by rhagstrand** [comment]
Made a new issue/question about the specific problem of getting the integrated version (the one found under packages/excalidraw, with a single javascript file under /dist to be used for integration in your own website) to work with a local excalidraw-rooms server here:

https://github.com/excalidraw/excalidraw/issues/9023

**Comment by prajwalagarwal5** [comment]
Hi @ad1992 Can you please confirm if this approach is still working. I tried everything mentioned but it doesn't seems to be working. If you can guide, it would be very helpfull. My use case is to integrate excalidraw in my existing video conference app

**Comment by mrkhachaturov** [comment]
<img width="2978" height="1604" alt="CleanShot 2025-12-24 at 21 09 12@2x" src="https://github.com/user-attachments/assets/7f7544b0-106a-4d56-9f2a-22cdff5c0d66" />
<img width="2956" height="1620" alt="CleanShot 2025-12-24 at 21 11 54@2x" src="https://github.com/user-attachments/assets/968cd59c-baea-49b2-98b4-05b6d313cec4" />


Hey everyone! For those interested in workspace functionality – I've been working on AstraDraw, a self-hosted Excalidraw fork that adds several enterprise features (screenshot attached):

Core Features:

- Workspaces – Personal and shared workspaces with role-based permissions (Admin/Member/Viewer)
- Authentication – OIDC integration + local auth with JWT-based sessions
- Collaboration – Real-time multi-user editing with presence indicators
- Comments System – Thread-based discussions anchored to canvas elements
- Presentation Mode – Navigate through canvas frames with a clean presenter view
- Talktrack – Record video explanations with screen capture and place them directly on the canvas as playable elements

Still refining the architecture and deployment process, but planning to share self-hosting guides and documentation soon. Will keep you posted! 🚀

**Reply by matbgn** [reply]
No github repo so far?

**Reply by mrkhachaturov** [reply]
> No github repo so far?

https://github.com/AstraDraw/astradraw
Still in early Alpha stage

**Comment by l4rm4nd** [comment]
I fixed it by adjusting the entrypoint.

See:

https://github.com/excalidraw/excalidraw/issues/5188#issuecomment-3848252703

https://blog.lrvt.de/selfhosting-excalidraw-with-collaboration-support/

**Comment by matbgn** [comment]
Yes and combined with non firebase storage would be delightful! 😍
